File: backend/app/models/tpm.py
from __future__ import annotations
import uuid
import enum
from sqlalchemy import (
    Column, String, Integer, Numeric, Boolean, Text, Date,
    ForeignKey, Enum, UniqueConstraint,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from app.db.base import Base, TimestampMixin


# ── Enums ─────────────────────────────────────────────────────────────────────

class TPMPeriodType(str, enum.Enum):
    ANNUAL       = "ANNUAL"
    QUARTERLY    = "QUARTERLY"
    MONTHLY      = "MONTHLY"
    EVENT_BASED  = "EVENT_BASED"


class TPMPlanStatus(str, enum.Enum):
    DRAFT        = "DRAFT"
    UNDER_REVIEW = "UNDER_REVIEW"
    APPROVED     = "APPROVED"
    ACTIVE       = "ACTIVE"
    CLOSED       = "CLOSED"
    ARCHIVED     = "ARCHIVED"


class TPMPromotionType(str, enum.Enum):
    DISCOUNT     = "DISCOUNT"
    FREE_GOODS   = "FREE_GOODS"
    VISIBILITY   = "VISIBILITY"
    DISPLAY      = "DISPLAY"
    REBATE       = "REBATE"
    OFF_INVOICE  = "OFF_INVOICE"
    BILL_BACK    = "BILL_BACK"
    LISTING_FEE  = "LISTING_FEE"
    BUNDLE       = "BUNDLE"
    EVENT        = "EVENT"
    CUSTOM       = "CUSTOM"


class TPMObjectiveType(str, enum.Enum):
    VOLUME              = "VOLUME"
    MARKET_SHARE        = "MARKET_SHARE"
    DISTRIBUTION_GAIN   = "DISTRIBUTION_GAIN"
    STOCK_CLEARANCE     = "STOCK_CLEARANCE"
    LAUNCH_SUPPORT      = "LAUNCH_SUPPORT"
    RETENTION           = "RETENTION"
    SEASONAL_PUSH       = "SEASONAL_PUSH"
    CHANNEL_ACTIVATION  = "CHANNEL_ACTIVATION"


class TPMPromotionStatus(str, enum.Enum):
    DRAFT      = "DRAFT"
    PROPOSED   = "PROPOSED"
    APPROVED   = "APPROVED"
    ACTIVE     = "ACTIVE"
    COMPLETED  = "COMPLETED"
    CANCELLED  = "CANCELLED"
    SETTLED    = "SETTLED"
    ARCHIVED   = "ARCHIVED"


class TPMBudgetType(str, enum.Enum):
    DISCOUNT_BUDGET    = "DISCOUNT_BUDGET"
    FREE_GOODS_BUDGET  = "FREE_GOODS_BUDGET"
    DISPLAY_BUDGET     = "DISPLAY_BUDGET"
    REBATE_BUDGET      = "REBATE_BUDGET"
    LISTING_FEE_BUDGET = "LISTING_FEE_BUDGET"
    MIXED              = "MIXED"


class TPMBaselineMethod(str, enum.Enum):
    PRIOR_PERIOD         = "PRIOR_PERIOD"
    SAME_PERIOD_LAST_YEAR = "SAME_PERIOD_LAST_YEAR"
    ROLLING_AVERAGE      = "ROLLING_AVERAGE"
    MANUAL               = "MANUAL"


class TPMClaimantType(str, enum.Enum):
    CUSTOMER    = "CUSTOMER"
    DISTRIBUTOR = "DISTRIBUTOR"
    INTERNAL    = "INTERNAL"


class TPMClaimType(str, enum.Enum):
    REBATE                      = "REBATE"
    DEDUCTION                   = "DEDUCTION"
    BILL_BACK                   = "BILL_BACK"
    DISPLAY_FEE                 = "DISPLAY_FEE"
    LISTING_FEE                 = "LISTING_FEE"
    FREE_GOODS_ADJUSTMENT       = "FREE_GOODS_ADJUSTMENT"
    OFF_INVOICE_RECONCILIATION  = "OFF_INVOICE_RECONCILIATION"
    CUSTOM                      = "CUSTOM"


class TPMClaimStatus(str, enum.Enum):
    DRAFT              = "DRAFT"
    SUBMITTED          = "SUBMITTED"
    UNDER_REVIEW       = "UNDER_REVIEW"
    APPROVED           = "APPROVED"
    PARTIALLY_SETTLED  = "PARTIALLY_SETTLED"
    SETTLED            = "SETTLED"
    REJECTED           = "REJECTED"
    CANCELLED          = "CANCELLED"


class TPMAIAgentType(str, enum.Enum):
    ROI_ANALYST         = "ROI_ANALYST"
    BUDGET_RISK_MONITOR = "BUDGET_RISK_MONITOR"
    PLANNER_ASSISTANT   = "PLANNER_ASSISTANT"


class TPMAIRecStatus(str, enum.Enum):
    PENDING      = "PENDING"
    ACKNOWLEDGED = "ACKNOWLEDGED"
    ACTIONED     = "ACTIONED"
    DISMISSED    = "DISMISSED"


# ── Models ────────────────────────────────────────────────────────────────────

class TPMPlan(Base, TimestampMixin):
    __tablename__ = "tpm_plans"

    id                   = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    plan_code            = Column(String(50), unique=True, nullable=False)
    plan_name            = Column(String(200), nullable=False)
    fiscal_year          = Column(Integer, nullable=False)
    period_type          = Column(Enum(TPMPeriodType), nullable=False, default=TPMPeriodType.ANNUAL)
    plan_start_date      = Column(Date, nullable=False)
    plan_end_date        = Column(Date, nullable=False)
    status               = Column(Enum(TPMPlanStatus), nullable=False, default=TPMPlanStatus.DRAFT)
    owner_user_id        = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    approved_by          = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    total_planned_budget = Column(Numeric(18, 2), default=0)
    total_approved_budget= Column(Numeric(18, 2), default=0)
    total_actual_spend   = Column(Numeric(18, 2), default=0)
    notes                = Column(Text)

    promotions = relationship("TPMPromotion", back_populates="plan", cascade="all, delete-orphan")


class TPMPromotion(Base, TimestampMixin):
    __tablename__ = "tpm_promotions"

    id                   = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    promotion_code       = Column(String(50), unique=True, nullable=False)
    promotion_name       = Column(String(200), nullable=False)
    tpm_plan_id          = Column(UUID(as_uuid=True), ForeignKey("tpm_plans.id"), nullable=True)
    linked_scheme_id     = Column(UUID(as_uuid=True), ForeignKey("promo_schemes.id"), nullable=True)
    promotion_type       = Column(Enum(TPMPromotionType), nullable=False)
    objective_type       = Column(Enum(TPMObjectiveType), nullable=False)
    status               = Column(Enum(TPMPromotionStatus), nullable=False, default=TPMPromotionStatus.DRAFT)
    valid_from           = Column(Date, nullable=False)
    valid_to             = Column(Date, nullable=False)
    brand_id             = Column(String(100))
    category_id          = Column(String(100))
    channel_id           = Column(String(100))
    region_id            = Column(String(100))
    customer_id          = Column(UUID(as_uuid=True), ForeignKey("customers.id"), nullable=True)
    distributor_group_id = Column(String(100))
    approved_by          = Column(UUID(as_uuid=True), ForeignKey("users.id"), nullable=True)
    notes                = Column(Text)

    plan          = relationship("TPMPlan", back_populates="promotions")
    budget_lines  = relationship("TPMBudgetLine", back_populates="promotion", cascade="all, delete-orphan")
    expected_perf = relationship("TPMExpectedPerf", back_populates="promotion", uselist=False, cascade="all, delete-orphan")
    actual_perf   = relationship("TPMActualPerf", back_populates="promotion", uselist=False, cascade="all, delete-orphan")
    claims        = relationship("TPMClaim", back_populates="promotion")


class TPMBudgetLine(Base, TimestampMixin):
    __tablename__ = "tpm_budget_lines"

    id                      = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tpm_promotion_id        = Column(UUID(as_uuid=True), ForeignKey("tpm_promotions.id"), nullable=False)
    budget_type             = Column(Enum(TPMBudgetType), nullable=False)
    planned_spend_amount    = Column(Numeric(18, 2), default=0)
    approved_spend_amount   = Column(Numeric(18, 2), default=0)
    actual_spend_amount     = Column(Numeric(18, 2), default=0)
    accrued_spend_amount    = Column(Numeric(18, 2), default=0)
    settled_spend_amount    = Column(Numeric(18, 2), default=0)
    remaining_budget_amount = Column(Numeric(18, 2), default=0)
    cost_center_id          = Column(UUID(as_uuid=True), ForeignKey("cost_centers.id"), nullable=True)
    dimension_value_id      = Column(UUID(as_uuid=True), ForeignKey("dim_values.id"), nullable=True)
    notes                   = Column(Text)

    promotion = relationship("TPMPromotion", back_populates="budget_lines")


class TPMExpectedPerf(Base, TimestampMixin):
    __tablename__ = "tpm_expected_perf"

    id                   = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tpm_promotion_id     = Column(UUID(as_uuid=True), ForeignKey("tpm_promotions.id"), nullable=False, unique=True)
    baseline_volume      = Column(Numeric(18, 3), default=0)
    target_volume        = Column(Numeric(18, 3), default=0)
    expected_uplift_qty  = Column(Numeric(18, 3), default=0)
    expected_uplift_pct  = Column(Numeric(10, 2), default=0)
    expected_revenue     = Column(Numeric(18, 2), default=0)
    expected_margin_impact = Column(Numeric(18, 2), default=0)
    expected_roi_pct     = Column(Numeric(10, 2), default=0)
    baseline_method      = Column(Enum(TPMBaselineMethod), nullable=False, default=TPMBaselineMethod.PRIOR_PERIOD)
    assumptions          = Column(Text)
    notes                = Column(Text)

    promotion = relationship("TPMPromotion", back_populates="expected_perf")


class TPMActualPerf(Base, TimestampMixin):
    __tablename__ = "tpm_actual_perf"

    id                   = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tpm_promotion_id     = Column(UUID(as_uuid=True), ForeignKey("tpm_promotions.id"), nullable=False, unique=True)
    actual_volume        = Column(Numeric(18, 3), default=0)
    actual_uplift_qty    = Column(Numeric(18, 3), default=0)
    actual_uplift_pct    = Column(Numeric(10, 2), default=0)
    actual_revenue       = Column(Numeric(18, 2), default=0)
    actual_margin_impact = Column(Numeric(18, 2), default=0)
    actual_spend         = Column(Numeric(18, 2), default=0)
    actual_roi_pct       = Column(Numeric(10, 2), default=0)
    post_event_notes     = Column(Text)

    promotion = relationship("TPMPromotion", back_populates="actual_perf")


class TPMClaim(Base, TimestampMixin):
    __tablename__ = "tpm_claims"

    id                    = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    claim_no              = Column(String(50), unique=True, nullable=False)
    tpm_promotion_id      = Column(UUID(as_uuid=True), ForeignKey("tpm_promotions.id"), nullable=False)
    claimant_type         = Column(Enum(TPMClaimantType), nullable=False)
    claimant_id           = Column(UUID(as_uuid=True), nullable=True)
    claim_date            = Column(Date, nullable=False)
    claim_type            = Column(Enum(TPMClaimType), nullable=False)
    claimed_amount        = Column(Numeric(18, 2), default=0)
    approved_amount       = Column(Numeric(18, 2), default=0)
    rejected_amount       = Column(Numeric(18, 2), default=0)
    settled_amount        = Column(Numeric(18, 2), default=0)
    status                = Column(Enum(TPMClaimStatus), nullable=False, default=TPMClaimStatus.DRAFT)
    reference_document_no = Column(String(100))
    reviewer_notes        = Column(Text)
    notes                 = Column(Text)

    promotion  = relationship("TPMPromotion", back_populates="claims")
    claim_lines = relationship("TPMClaimLine", back_populates="claim", cascade="all, delete-orphan")


class TPMClaimLine(Base, TimestampMixin):
    __tablename__ = "tpm_claim_lines"

    id                   = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tpm_claim_id         = Column(UUID(as_uuid=True), ForeignKey("tpm_claims.id"), nullable=False)
    source_order_id      = Column(UUID(as_uuid=True), ForeignKey("sales_orders.id"), nullable=True)
    source_scheme_id     = Column(UUID(as_uuid=True), ForeignKey("promo_schemes.id"), nullable=True)
    item_id              = Column(UUID(as_uuid=True), ForeignKey("products.id"), nullable=True)
    quantity_basis       = Column(Numeric(18, 3))
    rate_basis           = Column(Numeric(18, 4))
    claimed_amount       = Column(Numeric(18, 2), default=0)
    approved_amount      = Column(Numeric(18, 2), default=0)
    supporting_doc_ref   = Column(String(200))
    notes                = Column(Text)

    claim = relationship("TPMClaim", back_populates="claim_lines")


class TPMAIRecommendation(Base, TimestampMixin):
    __tablename__ = "tpm_ai_recommendations"

    id                = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    agent_type        = Column(Enum(TPMAIAgentType), nullable=False)
    tpm_promotion_id  = Column(UUID(as_uuid=True), ForeignKey("tpm_promotions.id"), nullable=True)
    tpm_claim_id      = Column(UUID(as_uuid=True), ForeignKey("tpm_claims.id"), nullable=True)
    title             = Column(String(300), nullable=False)
    detail            = Column(Text)
    severity          = Column(String(20))
    status            = Column(Enum(TPMAIRecStatus), nullable=False, default=TPMAIRecStatus.PENDING)
    actioned_notes    = Column(Text)

File: frontend/src/lib/tpm.ts
import axios from "axios";

const API = (process.env.NEXT_PUBLIC_API_URL ?? "http://localhost:8000") + "/api/v1";
const r = <T>(p: Promise<{ data: T }>) => p.then((res) => res.data);
const ax = axios.create({ baseURL: API });

// ── Enums & Types ──────────────────────────────────────────────────────────────

export type TPMPeriodType    = "ANNUAL" | "QUARTERLY" | "MONTHLY" | "EVENT_BASED";
export type TPMPlanStatus    = "DRAFT" | "UNDER_REVIEW" | "APPROVED" | "ACTIVE" | "CLOSED" | "ARCHIVED";
export type TPMPromotionType = "DISCOUNT" | "FREE_GOODS" | "VISIBILITY" | "DISPLAY" | "REBATE" | "OFF_INVOICE" | "BILL_BACK" | "LISTING_FEE" | "BUNDLE" | "EVENT" | "CUSTOM";
export type TPMObjectiveType = "VOLUME" | "MARKET_SHARE" | "DISTRIBUTION_GAIN" | "STOCK_CLEARANCE" | "LAUNCH_SUPPORT" | "RETENTION" | "SEASONAL_PUSH" | "CHANNEL_ACTIVATION";
export type TPMPromotionStatus = "DRAFT" | "PROPOSED" | "APPROVED" | "ACTIVE" | "COMPLETED" | "CANCELLED" | "SETTLED" | "ARCHIVED";
export type TPMBudgetType    = "DISCOUNT_BUDGET" | "FREE_GOODS_BUDGET" | "DISPLAY_BUDGET" | "REBATE_BUDGET" | "LISTING_FEE_BUDGET" | "MIXED";
export type TPMBaselineMethod = "PRIOR_PERIOD" | "SAME_PERIOD_LAST_YEAR" | "ROLLING_AVERAGE" | "MANUAL";
export type TPMClaimantType  = "CUSTOMER" | "DISTRIBUTOR" | "INTERNAL";
export type TPMClaimType     = "REBATE" | "DEDUCTION" | "BILL_BACK" | "DISPLAY_FEE" | "LISTING_FEE" | "FREE_GOODS_ADJUSTMENT" | "OFF_INVOICE_RECONCILIATION" | "CUSTOM";
export type TPMClaimStatus   = "DRAFT" | "SUBMITTED" | "UNDER_REVIEW" | "APPROVED" | "PARTIALLY_SETTLED" | "SETTLED" | "REJECTED" | "CANCELLED";
export type TPMAIAgentType   = "ROI_ANALYST" | "BUDGET_RISK_MONITOR" | "PLANNER_ASSISTANT";
export type TPMAIRecStatus   = "PENDING" | "ACKNOWLEDGED" | "ACTIONED" | "DISMISSED";

// ── Interfaces ─────────────────────────────────────────────────────────────────

export interface TPMBudgetLine {
  id: string;
  tpm_promotion_id: string;
  budget_type: TPMBudgetType;
  planned_spend_amount: number;
  approved_spend_amount: number;
  actual_spend_amount: number;
  accrued_spend_amount: number;
  settled_spend_amount: number;
  remaining_budget_amount: number;
  cost_center_id?: string;
  notes?: string;
}

export interface TPMExpectedPerf {
  id: string;
  tpm_promotion_id: string;
  baseline_volume: number;
  target_volume: number;
  expected_uplift_qty: number;
  expected_uplift_pct: number;
  expected_revenue: number;
  expected_margin_impact: number;
  expected_roi_pct: number;
  baseline_method: TPMBaselineMethod;
  assumptions?: string;
  notes?: string;
}

export interface TPMActualPerf {
  id: string;
  tpm_promotion_id: string;
  actual_volume: number;
  actual_uplift_qty: number;
  actual_uplift_pct: number;
  actual_revenue: number;
  actual_margin_impact: number;
  actual_spend: number;
  actual_roi_pct: number;
  post_event_notes?: string;
}

export interface TPMPromotion {
  id: string;
  promotion_code: string;
  promotion_name: string;
  tpm_plan_id?: string;
  linked_scheme_id?: string;
  promotion_type: TPMPromotionType;
  objective_type: TPMObjectiveType;
  status: TPMPromotionStatus;
  valid_from: string;
  valid_to: string;
  brand_id?: string;
  category_id?: string;
  channel_id?: string;
  region_id?: string;
  customer_id?: string;
  distributor_group_id?: string;
  notes?: string;
  created_at: string;
  budget_lines: TPMBudgetLine[];
  expected_perf?: TPMExpectedPerf;
  actual_perf?: TPMActualPerf;
  claims?: TPMClaim[];
}

export interface TPMPlan {
  id: string;
  plan_code: string;
  plan_name: string;
  fiscal_year: number;
  period_type: TPMPeriodType;
  plan_start_date: string;
  plan_end_date: string;
  status: TPMPlanStatus;
  total_planned_budget: number;
  total_approved_budget: number;
  total_actual_spend: number;
  notes?: string;
  created_at: string;
  promotion_count: number;
}

export interface TPMClaimLine {
  id: string;
  tpm_claim_id: string;
  source_order_id?: string;
  source_scheme_id?: string;
  item_id?: string;
  quantity_basis?: number;
  rate_basis?: number;
  claimed_amount: number;
  approved_amount: number;
  supporting_doc_ref?: string;
  notes?: string;
}

export interface TPMClaim {
  id: string;
  claim_no: string;
  tpm_promotion_id: string;
  claimant_type: TPMClaimantType;
  claimant_id?: string;
  claim_date: string;
  claim_type: TPMClaimType;
  claimed_amount: number;
  approved_amount: number;
  rejected_amount: number;
  settled_amount: number;
  status: TPMClaimStatus;
  reference_document_no?: string;
  reviewer_notes?: string;
  notes?: string;
  created_at: string;
  claim_lines: TPMClaimLine[];
}

export interface TPMAIRec {
  id: string;
  agent_type: TPMAIAgentType;
  tpm_promotion_id?: string;
  tpm_claim_id?: string;
  title: string;
  detail?: string;
  severity?: string;
  status: TPMAIRecStatus;
  actioned_notes?: string;
  created_at: string;
}

export interface TPMDashboard {
  active_plans: number;
  active_promotions: number;
  total_planned_budget: number;
  total_actual_spend: number;
  open_claims: number;
  open_claims_amount: number;
  pending_approvals: number;
  promotions_by_status: Record<string, number>;
  top_promotions: Array<{
    id: string;
    promotion_code: string;
    promotion_name: string;
    status: string;
    valid_from: string;
    valid_to: string;
  }>;
}

// ── Labels ─────────────────────────────────────────────────────────────────────

export const PERIOD_TYPE_LABEL: Record<TPMPeriodType, string> = {
  ANNUAL:      "Annual",
  QUARTERLY:   "Quarterly",
  MONTHLY:     "Monthly",
  EVENT_BASED: "Event-Based",
};

export const PLAN_STATUS_BADGE: Record<TPMPlanStatus, string> = {
  DRAFT:        "bg-gray-100 text-gray-600",
  UNDER_REVIEW: "bg-yellow-100 text-yellow-700",
  APPROVED:     "bg-blue-100 text-blue-700",
  ACTIVE:       "bg-green-100 text-green-700",
  CLOSED:       "bg-red-100 text-red-600",
  ARCHIVED:     "bg-gray-100 text-gray-500",
};

export const PROMOTION_TYPE_LABEL: Record<TPMPromotionType, string> = {
  DISCOUNT:    "Discount",
  FREE_GOODS:  "Free Goods",
  VISIBILITY:  "Visibility",
  DISPLAY:     "Display",
  REBATE:      "Rebate",
  OFF_INVOICE: "Off-Invoice",
  BILL_BACK:   "Bill-Back",
  LISTING_FEE: "Listing Fee",
  BUNDLE:      "Bundle",
  EVENT:       "Event",
  CUSTOM:      "Custom",
};

export const OBJECTIVE_TYPE_LABEL: Record<TPMObjectiveType, string> = {
  VOLUME:             "Volume Growth",
  MARKET_SHARE:       "Market Share",
  DISTRIBUTION_GAIN:  "Distribution Gain",
  STOCK_CLEARANCE:    "Stock Clearance",
  LAUNCH_SUPPORT:     "Launch Support",
  RETENTION:          "Retention",
  SEASONAL_PUSH:      "Seasonal Push",
  CHANNEL_ACTIVATION: "Channel Activation",
};

export const PROMOTION_STATUS_BADGE: Record<TPMPromotionStatus, string> = {
  DRAFT:     "bg-gray-100 text-gray-600",
  PROPOSED:  "bg-yellow-100 text-yellow-700",
  APPROVED:  "bg-blue-100 text-blue-700",
  ACTIVE:    "bg-green-100 text-green-700",
  COMPLETED: "bg-indigo-100 text-indigo-700",
  CANCELLED: "bg-red-100 text-red-600",
  SETTLED:   "bg-teal-100 text-teal-700",
  ARCHIVED:  "bg-gray-100 text-gray-500",
};

export const CLAIM_STATUS_BADGE: Record<TPMClaimStatus, string> = {
  DRAFT:             "bg-gray-100 text-gray-600",
  SUBMITTED:         "bg-blue-100 text-blue-700",
  UNDER_REVIEW:      "bg-yellow-100 text-yellow-700",
  APPROVED:          "bg-green-100 text-green-700",
  PARTIALLY_SETTLED: "bg-orange-100 text-orange-700",
  SETTLED:           "bg-teal-100 text-teal-700",
  REJECTED:          "bg-red-100 text-red-600",
  CANCELLED:         "bg-gray-100 text-gray-500",
};

export const AGENT_LABEL: Record<TPMAIAgentType, string> = {
  ROI_ANALYST:         "ROI Analyst",
  BUDGET_RISK_MONITOR: "Budget Risk Monitor",
  PLANNER_ASSISTANT:   "Planner Assistant",
};

export const SEVERITY_BADGE: Record<string, string> = {
  info:     "bg-blue-100 text-blue-700",
  warning:  "bg-yellow-100 text-yellow-700",
  critical: "bg-red-100 text-red-700",
};

export const fmtCurrency = (n: number) =>
  `KES ${Number(n).toLocaleString("en-KE", { minimumFractionDigits: 2 })}`;

// ── API Client ─────────────────────────────────────────────────────────────────

export const tpmApi = {
  // Dashboard
  getDashboard: () => r<TPMDashboard>(ax.get("/tpm/dashboard")),

  // Plans
  getPlans: (status?: TPMPlanStatus, fiscal_year?: number) =>
    r<TPMPlan[]>(ax.get("/tpm/plans", { params: { status, fiscal_year } })),
  createPlan: (data: object) => r<TPMPlan>(ax.post("/tpm/plans", data)),
  getPlan: (id: string) => r<TPMPlan>(ax.get(`/tpm/plans/${id}`)),
  approvePlan: (id: string) => r<TPMPlan>(ax.post(`/tpm/plans/${id}/approve`)),
  updatePlanStatus: (id: string, status: TPMPlanStatus) =>
    r<TPMPlan>(ax.patch(`/tpm/plans/${id}/status`, null, { params: { status } })),

  // Promotions
  getPromotions: (params?: { status?: string; plan_id?: string; promotion_type?: string; channel_id?: string; brand_id?: string }) =>
    r<TPMPromotion[]>(ax.get("/tpm/promotions", { params })),
  createPromotion: (data: object) => r<TPMPromotion>(ax.post("/tpm/promotions", data)),
  getPromotion: (id: string) => r<TPMPromotion>(ax.get(`/tpm/promotions/${id}`)),
  approvePromotion: (id: string) => r<TPMPromotion>(ax.post(`/tpm/promotions/${id}/approve`)),
  updatePromotionStatus: (id: string, status: TPMPromotionStatus) =>
    r<TPMPromotion>(ax.patch(`/tpm/promotions/${id}/status`, null, { params: { status } })),
  linkScheme: (id: string, scheme_id: string) =>
    r<TPMPromotion>(ax.post(`/tpm/promotions/${id}/link-scheme`, null, { params: { scheme_id } })),

  // Budget Lines
  addBudgetLine: (promo_id: string, data: object) =>
    r<TPMBudgetLine[]>(ax.post(`/tpm/promotions/${promo_id}/budget-lines`, data)),

  // Performance
  setExpectedPerf: (promo_id: string, data: object) =>
    r<TPMExpectedPerf>(ax.put(`/tpm/promotions/${promo_id}/expected-perf`, data)),
  setActualPerf: (promo_id: string, data: object) =>
    r<TPMActualPerf>(ax.put(`/tpm/promotions/${promo_id}/actual-perf`, data)),
  getPerformance: (promo_id: string) =>
    r<{ expected?: TPMExpectedPerf; actual?: TPMActualPerf }>(ax.get(`/tpm/promotions/${promo_id}/performance`)),

  // Claims
  getClaims: (params?: { status?: string; promotion_id?: string }) =>
    r<TPMClaim[]>(ax.get("/tpm/claims", { params })),
  createClaim: (data: object) => r<TPMClaim>(ax.post("/tpm/claims", data)),
  getClaim: (id: string) => r<TPMClaim>(ax.get(`/tpm/claims/${id}`)),
  reviewClaim: (id: string, data: object) => r<TPMClaim>(ax.post(`/tpm/claims/${id}/review`, data)),
  settleClaim: (id: string, data: object) => r<TPMClaim>(ax.post(`/tpm/claims/${id}/settle`, data)),

  // Reports
  getROIReport: () => r<any[]>(ax.get("/tpm/reports/roi")),
  getBudgetVsActual: (fiscal_year?: number) =>
    r<any[]>(ax.get("/tpm/reports/budget-vs-actual", { params: { fiscal_year } })),
  getClaimAging: () => r<any[]>(ax.get("/tpm/reports/claim-aging")),

  // AI
  runAgents: () => r<{ generated: number }>(ax.post("/tpm/ai/run-agents")),
  getAIRecs: (status?: TPMAIRecStatus) =>
    r<TPMAIRec[]>(ax.get("/tpm/ai/recommendations", { params: { status } })),
  ackAIRec: (id: string, data: object) =>
    r<TPMAIRec>(ax.patch(`/tpm/ai/recommendations/${id}`, data)),
};
